### ISSUE TITLE
VTM does not run on Android 15 with 16 KB page size

### ISSUE THREAD

**Issue by devemux86** [mapsforge/vtm#1125]
VTM does not run on 16 KB page size based Android 15.

_**It probably needs to rebuild the native libraries for Android.**_

- https://developer.android.com/guide/practices/page-sizes
- https://android-developers.googleblog.com/2024/08/adding-16-kb-page-size-to-android.html

**Comment by SammysHP** [comment]
Are you sure it doesn't work? For me it sounds like it's just an option to use 16 KB pages instead of 4 KB, but not a hard requirement.

**Comment by devemux86** [comment]
@SammysHP 

Set up the Android Emulator with a 16-KB-based, Android 15 system image and VTM does not work:
https://developer.android.com/guide/practices/page-sizes#16kb-emulator

(it works with regular Android 15 system image)

**Comment by devemux86** [comment]
We should see how we can improve [Mapsforge](https://github.com/mapsforge/mapsforge):

- https://github.com/mapsforge/mapsforge/issues/7
- https://github.com/mapsforge/mapsforge/issues/75

Then OpenGL is not needed anymore.

**Comment by devemux86** [comment]
We should see how we can improve [Mapsforge](https://github.com/mapsforge/mapsforge):

- https://github.com/mapsforge/mapsforge/issues/7

- https://github.com/mapsforge/mapsforge/issues/75

Then OpenGL will not be needed anymore.

**Comment by SammysHP** [comment]
Even if it was compatible, I just found this notice:

> We plan to make 16 KB page compatibility required for app uploads to the Google Play store next year.

(https://developer.android.com/about/versions/15/behavior-changes-all)

**Comment by devemux86** [comment]
I updated the build of the native libraries with #1126.

However, even if we rebuild the native libraries with Google's new guidelines,
it builds the same native libraries which do not run on Android 15 with 16 KB page size.

https://developer.android.com/guide/practices/page-sizes#compile-16-kb-alignment

If anyone knows better and can help, please try the build process.

**Comment by devemux86** [comment]
I updated the build of the native libraries with #1126.

However, even if we rebuild the native libraries with Android's new instructions,
it builds the same native libraries which do not run on Android 15 with 16 KB page size.

https://developer.android.com/guide/practices/page-sizes#compile-16-kb-alignment

---

If anyone knows better and can help, please try the build process:
- Set environment variable `NDK_HOME`
- Update ndk-build with [Android instructions](https://developer.android.com/guide/practices/page-sizes#compile-16-kb-alignment)
- Run `JniBuilder`

**Comment by devemux86** [comment]
I updated the process of building native libraries with #1126.

However, even if we rebuild the native libraries with [Android instructions](https://developer.android.com/guide/practices/page-sizes#compile-16-kb-alignment),
it builds the same native libraries which do not run on Android 15 with 16 KB page size.

---

If someone knows better, you can try to build the native libraries:
- Set environment variable `NDK_HOME`
- Update ndk-build with [Android instructions](https://developer.android.com/guide/practices/page-sizes#compile-16-kb-alignment)
- Run `JniBuilder`

**Comment by ocsike** [comment]
Hi,
Can you tried to build JNI with gdx-jnigen 2.5.2 ? 
https://github.com/libgdx/gdx-jnigen/pull/73

**Comment by devemux86** [comment]
@ocsike 

I no longer have a working environment for creating native libraries.

You could try building them and if they work correctly, I could merge them.
The instructions can be found above.

**Comment by ocsike** [comment]
I think it works with 16 KB....
1. I modified Android.mk and Application.mk then built libs with ndk-build. I modified JniBuilder.java too , but no need this case.
In Android.mk 
add APP_PLATFORM := android-35
add APP_SUPPORT_FLEXIBLE_PAGE_SIZES := true
and
modify LOCAL_LDLIBS := -lm -llog -Wl,-z,max-page-size=0x4000 
I added this, but I think above line is enough
LOCAL_LDFLAGS += -Wl,-z,max-page-size=16384
In Application.mk 
modify APP_PLATFORM := android-35



2. Copy these  libs into vtm-android\natives directory
3. Modify build.gradle  The androidCompileSdk() and androidTargetSdk() must be 35 
4. Build example app...
5. I use x86_64 emulator image with 16kb pagesize. In the emulator's config.ini must add this line as google suggested: 
kernel.parameters = androidboot.page_shift=14 
6. Start emulator ,install example app and start it
![vtm2](https://github.com/user-attachments/assets/dd1f2133-d259-42cc-ba30-f23c23886479)

**Comment by ztNFny** [comment]
@ocsike could you provide those 16k libs?

**Comment by ocsike** [comment]
> @ocsike could you provide those 16k libs?

@ztNFny 
You can download libs from here: szantai.info/vtmlibs.zip
This zip contains only .so files. I not tried physical device just only x86 emulator.

**Comment by devemux86** [comment]
@ocsike Thanks, I only make a quick test on Android emulator 16KB and they seem to work.

Do you want to post a pull request with the .so files and if changes are needed in the JNI module?
We only need .so files for `arm64-v8a`, `armeabi-v7a`, `x86`, `x86_64`.

**Comment by devemux86** [comment]
@ocsike Thanks, I made a quick test on Android 15 emulator 16KB and they seem to work.

But are they only made for Android 15? They do not work on older versions like **Android 5**.
Current .so files work on all Android versions.

**Comment by devemux86** [comment]
Probably need to use the existing [Application.mk](https://github.com/mapsforge/vtm/blob/master/jni/jni/Application.mk) with all ABI and platform an old Android.

**Comment by git-bone** [comment]
I built these with the existing Application.mk. I tested on emulators going from Lollipop to Baklava with 16kb page size. Seems to work fine on all of them with the android example app. 

Android MK:
```
LOCAL_PATH:= $(call my-dir)
include $(CLEAR_VARS)
 
LOCAL_MODULE    := vtm-jni
LOCAL_C_INCLUDES := . libtess2/Include 
 
LOCAL_CFLAGS := $(LOCAL_C_INCLUDES:%=-I%) -O2 -Wall -D__ANDROID__ -Wall -O2 -ffast-math -DNDEBUG
LOCAL_CPPFLAGS := $(LOCAL_C_INCLUDES:%=-I%) -O2 -Wall -D__ANDROID__ -Wall -O2 -ffast-math -DNDEBUG
LOCAL_LDLIBS := -lm -llog -Wl,-z,max-page-size=0x4000
LOCAL_LDFLAGS += "-Wl,-z,max-page-size=16384"
LOCAL_ARM_MODE  := arm
APP_SUPPORT_FLEXIBLE_PAGE_SIZES := true
 
LOCAL_SRC_FILES := libtess2/Source/dict.c\
	libtess2/Source/tess.c\
	libtess2/Source/geom.c\
	libtess2/Source/bucketalloc.c\
	libtess2/Source/mesh.c\
	libtess2/Source/sweep.c\
	libtess2/Source/priorityq.c\
	gl/utils.c\
	org_oscim_utils_TessJNI.cpp
 
include $(BUILD_SHARED_LIBRARY)
```


[vtmLibs.zip](https://github.com/user-attachments/files/18470075/vtmLibs.zip)

**Comment by devemux86** [comment]
@git-bone Thanks, these libs seem to work.

(I could only test them on Android x86 emulator)

**Comment by git-bone** [comment]
I tested on some old phones with android 9 and 14. Android 5 and 15 (16kb) on emulators. Everything worked fine.

**Comment by devemux86** [comment]
@git-bone 

@MotoUKRider reported this and I could reproduce it also on Android emulator with the mentioned map.

When the pull requests will be merged soon, the `snapshot` version on Maven and JitPack will include it too.

---

Technical info: the problem appears because we use the`mesh="true"` tesselation option in land and sea.

We better avoid this option, but Mapsforge / OpenAndroMaps / Freizeitkarte maps use [split land polygons](https://osmdata.openstreetmap.de/data/land-polygons.html)
that have empty grid space between the land tiles, more as we zoom in. (see image)
Mapsforge map engine can easily fill it with stroke, but VTM currently cannot do this.

If we find a way to fix this for VTM, then we can remove the `mesh="true"` tesselation rule.

![Image](https://github.com/user-attachments/assets/b977e349-e793-4231-9dd8-df5d54b604d5)

**Comment by devemux86** [comment]
@git-bone The triangle artifacts as they appear with the current native libraries north of Boston, UK:

<img src="https://github.com/user-attachments/assets/a566053a-ea76-41e8-b450-0cec60a0cab8" width="50%"/><img src="https://github.com/user-attachments/assets/db3d79e9-979b-43dc-8017-df157d3dea8f" width="50%"/>

**Comment by git-bone** [comment]
yes, i can reproduce it now. It only seems to happen on certain zoom levels. Never noticed this before.

**Comment by devemux86** [comment]
I published version `0.24.0` of VTM with the new Android native libraries on Maven and JitPack repositories.
